### PR TITLE
multi input layer

### PR DESCRIPTION
**Pull request by adam-r-kowalski**
This layer can be helpful in the case when your model takes in multiple inputs. For a Q network you take in both the state and the action and try to predict the expected sum of discounted rewards. Typically you would have to write this by constructing the different inputs separately.

```
state, action = randn(4), randn(2)

state_input = Dense(4, 20)
action_input = Dense(2, 20)
model = Chain(Dense(40, 20, elu), Dense(20, 1))
Q = model(vcat(state_input(state), action_input(action)))
```

However, this makes it awkward when trying to collect all the parameters or pass around the entire model as one struct.

Instead we can have a combinator which allows for multiple inputs and the whole implementation becomes simpler.

```
state, action = randn(4), randn(2)

model = Chain(MultiInput(Dense(4, 20), Dense(2, 20)),
              x -> vcat(x...),
              Dense(40, 20, elu),
              Dense(20, 1))

Q = model((state, action))
```

Now we have something which looks like a function again, and it just takes in multiple inputs. You can query for the parameters and make it part of a longer chain if needed.

### CODE REVIEW

**Comment by dhairyagandhi96** [comment]
I kinda like the sound of this. It got me thinking about if it could be extended it in Inception like situations, as well. The forward pass used currently is probably more expressive?

**Comment by chengchingwen** [comment]
OTOH, what's the different between this and #773. They're basically the same.

**Comment by Tokazama** [comment]
Thought he difference was multiple functions on one input versus multiple inputs with potentially different functions.

**Comment by adam-r-kowalski** [comment]
@Tokazama yup exactly. `Parallel` is when you wish to apply several functions to one input. `MultiInput` is when you wish to apply N functions to N inputs.

File: src/layers/basic.jl [hunk 1]
```diff
@@ -59,6 +59,46 @@ function activations(c::Chain, input)
...
+  MultiInput(xs...) = new{typeof(xs)}(xs)
+end
+
+@forward MultiInput.layers Base.getindex, Base.length, Base.first, Base.last,
+  Base.iterate, Base.lastindex
+
+children(m::MultiInput) = m.layers
+mapchildren(f, m::MultiInput) = MultiInput(f.(m.layers)...)
+
+(m::MultiInput)(xs) = [layer(x) for (layer, x) in zip(m.layers, xs)]
```
[the remark this answers is not in the document]
Comment (adam-r-kowalski) [reply src/layers/basic.jl]: I made the requested change! I'm surprised it doesn't lead to the same codegen?

[the remark this answers is not in the document]
Comment (chengchingwen) [reply src/layers/basic.jl]: It does what you request, here you request an array `[]`, so it give you an array. While map will try to remain the input type, so if you give it tuple, it give you tuple. And tuple can be unrolled by the compiler easily. check `@code_warntype` of two version.

**Comment by pshashk** [comment]
Maybe we can combine them into one and choose based on the input tuple size?
```julia
Unchain(layers...)(inputs...) # equivalent to MultiInput
Unchain(layers...)((input,)) # equivalent to Parallel without reduction
```

**Comment by Tokazama** [comment]
@pshashk , I think that's what bogged down the previous proposal. It seems intuitive that this sort of thing be rolled into one Type, but it's hard to account for all the ways it may be used. For example, what if I want to apply a set of different functions to to a tuple of inputs.

```julia
Parallel((x, y, z)) --> ((x1,y1,z1), (x2,y2,z2))
```
The previous strategy proposed was to have a function that mapped the input to the different parallel chains. But then you have to decide what is a sane default and design for every use case.

IMO. Unless someone comes up with a very intuitive and simple way of designing novel network topology through a single layer, it makes more sense to have distinct separate layers that can freely be combined.

**Comment by pshashk** [comment]
Yes, Parallel with tuple input looks overly verbose:
```julia 
Chain(x->(x,), Unchain(layers...))
```
What about making MultiInput look like a broadcasted Parallel?
```julia
Unchain(layers...)(input) # Parallel
Unchain(layers...).(inputs) # MultiInput
```

**Comment by adam-r-kowalski** [comment]
I agree, ideally we would only have one layer instead of both of these. I think I might have a solution. What if we only have one layer called `Parallel` and it works like `MultiInput` currently does. However, when you want to apply multiple functions to one input you don't want to write `Parallel(f1, f2, f3)(x, x, x)` So what we do is have a new layer called `FanOut` which can repeat your input `n` times. What does this mean in practice?

If you wish to have the current behavior of applying N functions to 1 input then you can write
```
model = Chain(FanOut(3), Parallel(f1, f2, f3))
x = randn(4)
model(input) == [f1(x), f2(x), f3(x)]
```

However, if you wish to apply N functions t N inputs then you can write
```
model = Parallel(f1, f2, f3)
x1, x2, x3 = randn(4), randn(2), randn(3)
model((x1, x2, x3)) == [f1(x1), f2(x2), f3(x3)
```

This way the control flow is simple and obvious. The precedent from this is again the Jax library, and it seems to work quite well for them.

**Comment by kczimm** [comment]
Is there any update on this work? I’m interested in implementing [UNET](https://arxiv.org/abs/1505.04597) and believe I need something like this type of layer.

**Comment by pshashk** [comment]
You can do UNET without Parallel layer. Downsampling path is basically `accumulate((x, layer) -> layer(x), layers, init = input)` and upsampling can be done similarly via `foldr` where `init` is the bottleneck and `itr` is zipped upsampling layers and intermediate results from the down path.

**Comment by kczimm** [comment]
@pshashk. This is great thanks!

**Comment by bhvieira** [comment]
> 
> 
> Is there any update on this work? I’m interested in implementing [UNET](https://arxiv.org/abs/1505.04597) and believe I need something like this type of layer.

You only need the `Flux.SkipConnection` for the U-Net.

**Comment by arnavs** [comment]
What's the current status of this sort of thing? Does the functionality exist in Flux?

**Comment by bionicles** [comment]
it's OK to have multiple layers doing similar-ish things because functions are cattle, not pets. Also, users can implement their own stuff easily. [Look at Ramda](https://ramdajs.com/docs/). Many functions are quite similar, juxt, converge, useWith. Chain is Pipe. Super useful

```
function MultiInput(fns)
    function F(xs)
            return (f(x) for (f, x) in zip(fns, xs))
     end
     return F
end

function Parallel(fns)
    function F(x)
           return (f(x) for f in fns)
     end
end
```

However, these layers which produce multiple outputs are gonna always have some merge op afterwards, so perhaps we can just include them?

```
function MultiInputWithCat(fns; cat_fn=vcat)
    function F(xs)
          return cat_fn((f(x) for (f,x) in zip(fns, xs))...)
    end
end
```
Perhaps, we ought to just make a Flux Addons and build special layers and helpers there...flux can just be core stuff

**Comment by ToucheSir** [comment]
Thanks for the PR! It's taken a while, but I think we finally have something pretty general-purporse with the new `Parallel` layer. e.g. the example up top would look like:
```julia
model = Chain(Parallel(vcat, Dense(4, 20), Dense(2, 20)),
              Dense(40, 20, elu),
              Dense(20, 1))
```
So if there are no objections, this should be safe to close.

**Pull request closed without merging**